src/pint/models/model_builder.py

```diff
@@ -26,276 +26,296 @@ class UnknownBinaryModel(ValueError):
-                )
-                prefix_in_model = self.timing_model.get_params_of_type_top(ptype)
-                for key in prefix_param:
-                    ppnames = [x for x in prefix_in_model if x.startswith(key)]
-                    for ppn in ppnames:
-                        pfx, idxs, idxv = split_prefixed_name(ppn)
-                        if pfx == key:
-                            exm_par = getattr(self.timing_model, ppn)
-                        else:
-                            continue
-                    exm_par_comp = param_inModel[exm_par.name]
-                    for parname in prefix_param[key]:
-                        pre, idstr, idx = split_prefixed_name(parname)
-                        if idx == exm_par.index:
-                            continue
-                        if hasattr(exm_par, "new_param"):
-                            new_par = exm_par.new_param(idx)
-                            self.timing_model.add_param_from_top(new_par, exm_par_comp)
-            if "BINARY" in self.param_inparF:
-                vals = self.param_inparF["BINARY"]
-                if len(vals) != 1:
-                    raise ValueError(
-                        "Mal-formed binary model selection: {}".format(
-                            repr(" ".join(["BINARY"] + vals))
-                        )
-                    )
-                (bm,) = vals
-                cats = self.timing_model.get_components_by_category()
-                if "pulsar_system" not in cats:
-                    raise UnknownBinaryModel(
-                        "Unknown binary model requested in par file: {}".format(bm)
-                    )
-                # FIXME: consistency check - the componens actually chosen should know the name bm
-        for p in self.timing_model.params:
-            if isinstance(self.timing_model[p], maskParameter):
-                # maskParameters need a bogus alias for parfile parsing
-                # remove this bogus alias
-                try:
-                    ix = self.timing_model[p].aliases.index(self.timing_model[p].prefix)
-                except ValueError:
-                    pass
-                else:
-                    del self.timing_model[p].aliases[ix]
-        if parfile is not None:
-            self.timing_model.read_parfile(parfile)
-
-    def get_control_info(self):
-        info = {}
-        if not self.param_unrecognized == {}:
-            for ctrlp in self.control_params:
-                if ctrlp in self.param_unrecognized:
-                    info[ctrlp] = self.param_unrecognized[ctrlp]
+        result_componets = []
+        param_inpar = self.parse_parfile(parfile)
+        # 1. iteration read parfile with a no component timing_model to get
+        # the overall control parameters. This will get us the binary model name
+        # build the base fo the timing model
+        self.timing_model.read_parfile(parfile)
```

Comment:
Can't you just look up BINARY in `param_inpar`?

Reply:
Changed.